src/mbgl/style/layers/custom_drawable_layer.cpp

```diff
@@ -274,15 +336,98 @@ void CustomDrawableLayerHost::Interface::addFill(const GeometryCollection& geome
+        builder = createBuilder("symbol", symbolShader);
+    }
+    assert(builder);
+    assert(builder->getShader() == symbolShader);
+
+    // temporary: buffers
+    struct CustomSymbolIcon {
+        std::array<float, 2> a_pos;
+        std::array<float, 2> a_tex;
+    };
+
+    // vertices
+    using VertexVector = gfx::VertexVector<CustomSymbolIcon>;
+    const std::shared_ptr<VertexVector> sharedVertices = std::make_shared<VertexVector>();
+    VertexVector& vertices = *sharedVertices;
+
+    // encode center and extrude direction into vertices
+    for (int y = 0; y <= 1; ++y) {
+        for (int x = 0; x <= 1; ++x) {
+            vertices.emplace_back(
+                CustomSymbolIcon{{static_cast<float>(point.x * 2 + x), static_cast<float>(point.y * 2 + y)},
+                                 {symbolOptions.textureCoordinates[x][0], symbolOptions.textureCoordinates[y][1]}});
+        }
+    }
+
+    // indexes
+    using TriangleIndexVector = gfx::IndexVector<gfx::Triangles>;
+    const std::shared_ptr<TriangleIndexVector> sharedTriangles = std::make_shared<TriangleIndexVector>();
+    TriangleIndexVector& triangles = *sharedTriangles;
+
+    triangles.emplace_back(0, 1, 2, 1, 2, 3);
+
+    SegmentVector<CustomSymbolIcon> triangleSegments;
+    triangleSegments.emplace_back(Segment<CustomSymbolIcon>{0, 0, 4, 6});
+
+    // add to builder
+    static const StringIdentity idPositionAttribName = stringIndexer().get("a_pos");
+    static const StringIdentity idTextureAttribName = stringIndexer().get("a_tex");
+    builder->setVertexAttrNameId(idPositionAttribName);
+
+    auto attrs = context.createVertexAttributeArray();
+    if (const auto& attr = attrs->add(idPositionAttribName)) {
+        attr->setSharedRawData(sharedVertices,
+                               offsetof(CustomSymbolIcon, a_pos),
+                               /*vertexOffset=*/0,
+                               sizeof(CustomSymbolIcon),
+                               gfx::AttributeDataType::Float2);
+    }
+    if (const auto& attr = attrs->add(idTextureAttribName)) {
+        attr->setSharedRawData(sharedVertices,
+                               offsetof(CustomSymbolIcon, a_tex),
+                               /*vertexOffset=*/0,
+                               sizeof(CustomSymbolIcon),
+                               gfx::AttributeDataType::Float2);
+    }
+    builder->setVertexAttributes(std::move(attrs));
+    builder->setRawVertices({}, vertices.elements(), gfx::AttributeDataType::Float2);
+    builder->setSegments(gfx::Triangles(), sharedTriangles, triangleSegments.data(), triangleSegments.size());
+
+    // texture
+    if (symbolOptions.texture) {
+        static const StringIdentity idTextureUniformName = stringIndexer().get("u_texture");
+        if (auto textureLocation = builder->getShader()->getSamplerLocation(idTextureUniformName);
+            textureLocation.has_value()) {
+            builder->setTexture(symbolOptions.texture, *textureLocation);
```

Comment:
I just merged https://github.com/maplibre/maplibre-native/pull/2107, so you probably need to update this code in order to compile.